flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/fetcher/KafkaSourceFetcherManager.java

```diff
@@ -73,7 +73,7 @@ public void commitOffsets(
         if (offsetsToCommit.isEmpty()) {
             return;
         }
-        SplitFetcher<Tuple3<T, Long, Long>, KafkaPartitionSplit> splitFetcher = fetchers.get(0);
+        SplitFetcher<Tuple3<T, Long, Long>, KafkaPartitionSplit> splitFetcher = getRunningFetcher();
```

Comment:
Nit: Does this change have any effect on the fix if not maybe make the change a separate commit?

Reply:
Can you explain a bit more why this is a performance improvement?

Reply:
Thanks for the review @fapaul. This change does not affect the fix. I have updated the PR to remove this change.

Regarding the reason why this could improve performance, let's assume the first fetcher created by this `KafkaSourceFetcherManager` has been closed and removed from `fetchers`. Prior to this change, every time the `commitOffsets()` is called, it will create a new `SplitFetcher` just to commit the offset. If `commitOffsets()` is called N times, then N `SplitFetcher` will be created, which seems to be really inefficient.

In order to fix this problem, we can commit the message using any running fetcher in the `fetchers`, which could be achieved by using `getRunningFetcher()` here.

Reply:
I created https://issues.apache.org/jira/browse/FLINK-24398 to track this issue.